cryptoki/src/context/mod.rs

```diff
@@ -32,18 +33,50 @@ use derivative::Derivative;
 use log::error;
 use std::mem;
 use std::path::Path;
+use std::ptr;
+use std::sync::Arc;
 
-/// Main PKCS11 context. Should usually be unique per application.
 #[derive(Derivative)]
 #[derivative(Debug)]
-pub struct Pkcs11 {
+// Implementation of Pkcs11 class that can be enclosed in a single Arc
+pub(crate) struct Pkcs11Impl {
     // Even if this field is never read, it is needed for the pointers in function_list to remain
     // valid.
     #[derivative(Debug = "ignore")]
     _pkcs11_lib: cryptoki_sys::Pkcs11,
     pub(crate) function_list: cryptoki_sys::_CK_FUNCTION_LIST,
 }
 
+impl Pkcs11Impl {
+    // Private finalize call
+    #[inline(always)]
+    fn finalize(&self) -> Result<()> {
+        unsafe {
+            Rv::from(self
+                .function_list
+                .C_Finalize
+                .ok_or(Error::NullFunctionPointer)?(
+                ptr::null_mut()
+            ))
+            .into_result()
+        }
+    }
+}
+
+impl Drop for Pkcs11Impl {
+    fn drop(&mut self) {
+        if let Err(e) = self.finalize() {
+            error!("Failed to finalize: {}", e);
+        }
+    }
+}
+
+/// Main PKCS11 context. Should usually be unique per application.
+#[derive(Clone, Debug)]
```

Comment:
Is it needed/safe to make this clonable? I think we only have methods that require a non-mutable reference, so one instance could be used across threads. For creating sessions we could just build another `Pkcs11` on the fly.

Reply:
I believe it is safe to make `Pkcs11` clonable because it will clone the instance of `Pkcs11Impl`.  I dont' think you want to make a new `Pkcs11` per slot as that will attempt to reload the PKCS11 C library and call `C_GetFunctionList` again.

Reply:
Hmm, true. I mean `Pkcs11` can be used across threads safely, I assume, but cloning shouldn't be that bad since we'd still have one "actual" context per app.